### DISCUSSION TITLE
How to configure fluent nhibernate unicode support ?

### DISCUSSION THREAD

**Discussion by HakanKaraoglu** [nhibernate/fluent-nhibernate#506] [Q&A] [answered]
I am sending unicode characters (e.g. emoji) from the application to the database but they are "?" seen as. I used the following configuration, but I did not get any results, can anyone who knows about this help?

```
 var configuration = Fluently.Configure()
                .Database(OracleManagedDataClientConfiguration.Oracle10
                .ConnectionString(connectionString)                            
                .Cache(x=>x.ProviderClass<CoreMemoryCacheProvider>().UseQueryCache().UseSecondLevelCache().UseMinimalPuts())
                .ExposeConfiguration(c => 
                    c.SetProperty("generate_statistics", "true")
                     .SetProperty("useUnicode", "true")
                     .SetProperty("defaultNChar", "true")
                     .SetProperty("characterEncoding", "utf-8")
                )
                .Mappings(m => m.FluentMappings.AddFromAssemblyOf<ISniffactorMapping>().Conventions.Add(ConventionBuilder.Class.Always(x=>x.Cache.ReadWrite())));
```

**Comment by fredericDelaporte** [comment, marked as answer]
There is very likely a mismatch in encoding anywhere between how the data is input, writen to the database (check the schema encoding there), read again from database, then output in your application.
So it is hard to tell where things are going wrong.

General explanation and guidance on such encoding troubles can be found [here](https://kunststube.net/encoding/), just in case.

Check also your Oracle configuration. Oracle Unicode handling may be tricky, depending on how it is configured. So you have to be sure you have set it adequately and use matching string types, types which vary according to the configuration of your server. And also NHibernate Oracle dialect configuration may have to be adjusted depending on how the server is configured.
See the `oracle.use_n_prefixed_types_for_unicode` setting in [optional configuration properties documentation](https://nhibernate.info/doc/nhibernate-reference/session-configuration.html#configuration-optional-dialects).

Anyway NHibernate has a few tests for writing non ascii characters, runs them on Oracle, and has no trouble with them. See by example [here](https://github.com/nhibernate/nhibernate-core/blob/ae69a0686954c599232fac0435ed1361012dd32f/src/NHibernate.Test/TypesTest/StringTypeFixture.cs#L44-L59). If you can provide a failing test in case the trouble is actually on NHibernate side, we could have a look into it.
